### PR TITLE
Improve activation function tests

### PR DESCRIPTION
**Pull request by okuta**
This PR stabilizes activation tests.
Please merge #3330 before this PR.

- Fix random seed
- Adjust threshold
- Simplify code

### CODE REVIEW

File: tests/chainer_tests/functions_tests/activation_tests/test_crelu.py [hunk 1]
```diff
@@ -8,15 +8,14 @@
...
 
 def _replace_near_zero_values(x):
     # Replace near zero values in an array in order to avoid unstability of
     # numerical grad
-    for i in range(x.size):
-        if -0.01 < x.flat[i] < 0.01:
-            x.flat[i] = 0.5
+    for i in numpy.ndindex(x.shape):
+        if -0.01 < x[i] < 0.01:
+            x[i] = 0.5
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/activation_tests/test_crelu.py]: How about: `x[numpy.logical_and(-0.01 < x, x < 0.01)] = 0.5` ?

File: tests/chainer_tests/functions_tests/activation_tests/test_hard_sigmoid.py [hunk 2]
```diff
@@ -26,40 +26,32 @@ def setUp(self):
...
     @attr.gpu
-    @condition.retry(3)
     def test_forward_gpu(self):
         self.check_forward(cuda.to_gpu(self.x))
 
     def check_backward(self, x_data, grad):
         gradient_check.check_backward(
-            functions.HardSigmoid(), x_data, grad,
-            **self.check_backward_options)
+            functions.HardSigmoid(), x_data, grad, dtype=numpy.float64)
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/activation_tests/test_hard_sigmoid.py]: This test seems unstable.
Can you adjust tolerance?

File: tests/chainer_tests/functions_tests/activation_tests/test_leaky_relu.py [hunk 3]
```diff
@@ -28,31 +28,27 @@ def setUp(self):
...
         self.check_forward_options = {}
-        self.check_backward_options = {'dtype': numpy.float64}
+        self.check_backward_options = {}
+        self.check_double_backward_options = {}
         if self.dtype == numpy.float16:
             self.check_forward_options = {'atol': 1e-4, 'rtol': 1e-3}
-            self.check_backward_options = {
-                'dtype': numpy.float64, 'atol': 5e-3, 'rtol': 5e-2}
+            self.check_backward_options = {'atol': 1e-4, 'rtol': 1e-3}
+            self.check_double_backward_options = {'atol': 5e-4, 'rtol': 5e-3}
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/activation_tests/test_leaky_relu.py]: In my environment, this configuration is too strict (both backward and double-backward).
Can you relax it?

File: tests/chainer_tests/functions_tests/activation_tests/test_crelu.py (hunk 1 above)
Comment (okuta) [reply tests/chainer_tests/functions_tests/activation_tests/test_crelu.py]: Good!

File: tests/chainer_tests/functions_tests/activation_tests/test_log_softmax.py [hunk 4]
```diff
@@ -29,11 +29,12 @@ def setUp(self):
...
         self.check_forward_options = {}
-        self.check_backward_options = {'dtype': numpy.float64}
+        self.check_backward_options = {}
+        self.check_double_backward_options = {}
         if self.dtype == numpy.float16:
             self.check_forward_options = {'atol': 5e-3, 'rtol': 5e-2}
-            self.check_backward_options = {
-                'dtype': numpy.float64, 'atol': 5e-4, 'rtol': 5e-3}
+            self.check_backward_options = {'atol': 5e-4, 'rtol': 5e-3}
+            self.check_double_backward_options = {'atol': 5e-3, 'rtol': 5e-2}
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/activation_tests/test_log_softmax.py]: In my environment, backward (not double-backward) test on float16 is unstable.
Could you relax it?

**Comment by niboshi** [comment]
Here is a list of tests that are unstable (all float16):

* `test_leaky_relu` - backward
* `test_log_softmax` - backward
* `test_lstm` - backward
* `test_sigmoid` - forward, backward
* `test_slstm` - forward, backward
* `test_softplus` - forward
* `test_tanh` - double-backward, backward

**Comment by niboshi** [comment]
These tests are still unstable (all float16):
* `test_leaky_relu` - double-backward (not backward)
   - this is relatively rare: failed after 551 iterations.
* `test_log_softmax` - backward
* `test_lstm` - backward
* `test_slstm` - backwkard

**Comment by okuta** [comment]
I resolved conflict and fixed.

**Comment by niboshi** [comment]
jenkins, test this please

**Comment by niboshi** [comment]
LGTM!

**Pull request merged**